worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,887 @@
+	use substratee_worker_primitives::{
+		block::{Block, SignedBlock as SignedSidechainBlock},
+		traits::{
+			Block as SidechainBlockTrait, SignBlock, SignedBlock as SignedSidechainBlockTrait,
+		},
+	};
+
+	#[test]
+	fn load_shards_from_db_works() {
+		// given
+		let path = PathBuf::from("load_shards_from_db_works");
+		let shard_one = H256::from_low_u64_be(1);
+		let shard_two = H256::from_low_u64_be(2);
+		// when
+		{
+			let sidechain_db = SidechainStorage::new(path.clone()).unwrap();
+			// ensure db starts empty
+			assert_eq!(SidechainStorage::load_shards_from_db(&sidechain_db.db).unwrap(), vec![]);
+			// write signed_block to db
+			sidechain_db
+				.db
+				.put(STORED_SHARDS_KEY.encode(), vec![shard_one, shard_two].encode())
+				.unwrap();
+		}
+
+		// then
+		{
+			// open new DB of same path:
+			let updated_sidechain_db = SidechainStorage::new(path.clone()).unwrap();
+			let loaded_shards =
+				SidechainStorage::load_shards_from_db(&updated_sidechain_db.db).unwrap();
+			assert!(loaded_shards.contains(&shard_one));
+			assert!(loaded_shards.contains(&shard_two));
+		}
+		// clean up
+		let _ = DB::destroy(&Options::default(), path).unwrap();
+	}
+
+	#[test]
+	fn load_last_block_from_db_works() {
+		// given
+		let path = PathBuf::from("load_last_block_from_db_works");
+		let shard = H256::from_low_u64_be(1);
+		let signed_block = create_signed_block(20, shard);
+		let signed_last_block = LastSidechainBlock {
+			hash: signed_block.hash(),
+			number: signed_block.block().block_number(),
+		};
+		// when
+		{
+			let sidechain_db = SidechainStorage::new(path.clone()).unwrap();
+			// ensure db starts empty
+			assert!(SidechainStorage::load_last_block_from_db(&sidechain_db.db, &shard)
+				.unwrap()
+				.is_none());
+			// write signed_block to db
+			sidechain_db
+				.db
+				.put((LAST_BLOCK_KEY, shard).encode(), signed_last_block.encode())
+				.unwrap();
+		}
+
+		// then
+		{
+			// open new DB of same path:
+			let updated_sidechain_db = SidechainStorage::new(path.clone()).unwrap();
+			let loaded_block =
+				SidechainStorage::load_last_block_from_db(&updated_sidechain_db.db, &shard)
+					.unwrap()
+					.unwrap();
+			assert_eq!(loaded_block, signed_last_block);
+		}
+		// clean up
+		let _ = DB::destroy(&Options::default(), path).unwrap();
+	}
+
+	#[test]
+	fn create_new_sidechain_storage_works() {
+		// given
+		let path = PathBuf::from("create_new_sidechain_storage_works");
+		let shard = H256::from_low_u64_be(1);
+		let shard_vector = vec![shard];
+		let signed_block = create_signed_block(20, shard);
+		let signed_last_block = LastSidechainBlock {
+			hash: signed_block.hash(),
+			number: signed_block.block().block_number(),
+		};
+		// when
+		{
+			let sidechain_db = SidechainStorage::new(path.clone()).unwrap();
+			// ensure db starts empty
+			assert!(SidechainStorage::load_last_block_from_db(&sidechain_db.db, &shard)
+				.unwrap()
+				.is_none());
+			// write shards to db
+			sidechain_db
```

Comment:
If I look at this, I'd suggest a wrapper function:

```
fn put<K: Encode, V: Encode>(key: K, value: V) ....
```

So we don't need to always do encode when we add something.

Reply:
Added with commit https://github.com/integritee-network/worker/pull/380/commits/fdf65a90e6c158f18894b018a77703a4902c7b3d

Does that reach your bar? :)